src/libcore/iter/adapters/mod.rs

```diff
@@ -17,6 +18,40 @@ pub use self::flatten::{FlatMap, Flatten};
 pub use self::zip::Zip;
 pub(crate) use self::zip::TrustedRandomAccess;
 
+/// This trait provides access to to the backing source of an interator-adapter pipeline
+/// under the conditions that
+/// * the iterator source `S` itself implements `SourceIter<Source = S>`
+/// * there is a delegating implementation of this trait for each adapter in the pipeline
+///
+/// This is useful for specializing [`FromIterator`] implementations or to retrieve
+/// the remaining values from a source of a partially consumed iterator.
+///
+/// # Examples
+///
+/// Retrieving a partially consumed source and wrapping it into a different pipeline:
+///
+/// ```
+/// # #![feature(inplace_iteration)]
+/// # use std::iter::SourceIter;
+///
+/// let mut iter = vec![9, 9, 9].into_iter().map(|i| i * i);
+/// let first = iter.next().unwrap();
+/// let mut remainder = std::mem::replace(iter.as_inner(), Vec::new().into_iter());
+/// let second = remainder.map(|i| i + 1).next().unwrap();
+/// assert_eq!(first, 81);
+/// assert_eq!(second, 10);
```

Comment:
I realized that providing access to the underlying source exposes a lot of implementation details of pipeline stages.

I'll update the documentation to warn users that while this trait provides access it will make very few guarantees about the state of the underlying source after it has been in the hands of adapters with unspecified internal behavior.

Reply:
I'm not totally convinced that documentation will be good enough.

Reply:
The VecDeque implementation is the only real problematic one I see I think. That should just return the vec_deque::IntoIter.

Reply:
`vec_deque::IntoIter` doesn't opt into this.

But you're right nevertheless that as-is it already allows unsafe things to happen. An adapter could swap out its underlying source in the middle of iteration, thus violating the assumptions of the Vec specialization.

I'll mark the trait as unsafe then and add asserts to the specialization.